app/routes.py

```diff
@@ -1,2 +1,239 @@
+
+    task.completed_at = datetime.utcnow()
+    db.session.commit()
+    
+    slack_bot(task.title)
+        
+    response_body = {"task": 
+        {"id": task.id,
+        "title": task.title,
+        "description": task.description,
+        "is_complete": bool(task.completed_at)}}
+    return jsonify(response_body), 200
+
+
+@tasks_bp.route("/<task_id>/mark_incomplete", methods=["PATCH"])
+def handle_incompleted_task(task_id):
+    task = Task.query.get_or_404(task_id)
+    task.completed_at = None
+    db.session.commit()
+        
+    response_body = {"task": 
+        {"id": task.id,
+        "title": task.title,
+        "description": task.description,
+        "is_complete": bool(task.completed_at)}}
+    
+    return jsonify(response_body), 200
+    
+
+
+
+goals_bp = Blueprint("goals", __name__, url_prefix="/goals")
+
+@goals_bp.route("", methods=["POST"])
+def post_goals():
+    request_body = request.get_json()
+    if "title" not in request_body:
+        return make_response({"details":"Invalid data"}, 400)
+    
+    new_goal = Goal(
+        title = request_body["title"],
+        )
+
+    db.session.add(new_goal)
+    db.session.commit()
+    
+    return {"goal": {"id": new_goal.id,
+        "title": new_goal.title}}, 201
+
+
+@goals_bp.route("", methods=["GET"])   
+def get_goals():
+        goals = Goal.query.all()
+    
+        goals_response = []
+        for goal in goals: 
+            goals_response.append({
+                "id": goal.id,
+                "title": goal.title
+                })
+
+        return jsonify(goals_response), 200
+
+
+@goals_bp.route("/<goal_id>", methods=["GET", "PUT", "DELETE"])
+
+def handle_goal(goal_id):
+    goal = Goal.query.get_or_404(goal_id)
+    if not goal:
+        return make_response(f"Goal {goal_id} not found", 404)
+        
+    if request.method == "GET":
+        response_body = {"goal":
+            {"id": goal.id,
+            "title": goal.title,
+            }}
+        return jsonify(response_body), 200
+
+    elif request.method == "PUT":
+        request_body = request.get_json()
+        goal.title = request_body['title']
+        db.session.commit()
+        
+        response_body = {"goal": {"id": goal.id,
+            "title": goal.title,
+            }}
+        return jsonify(response_body), 200
+
+    elif request.method == "DELETE":
+        db.session.delete(goal)
+        db.session.commit()
+
+        return {"details":(f'Goal {goal.id} "{goal.title}" successfully deleted')}, 200
+
+
+@goals_bp.route("/<goal_id>/tasks", methods=["GET", "POST"])
+def handle_goals_tasks(goal_id):
+    goal = Goal.query.get_or_404(goal_id)
+    if not goal: 
+        return make_response ("Goal not found", 404)
+
+    if request.method == "POST":
+        request_body = request.get_json()
+
+        all_ids = []
+        for task_id in request_body['task_ids']: 
+            all_ids.append(Task.query.get(task_id))
+        goal.tasks = all_ids
```

Comment:
👍 

Assigning to the `tasks` relationship is absolutely a valid way to associate tasks with a goal.  We should also think about whether to interpret POSTing to this endpoint is setting the exact list of tasks that should be a part of the goal, or whether we are adding these tasks to the goal (in addition to anything that was already there). Neither the project description nor the test `test_post_task_ids_to_goal_already_with_goals` are entirely clear on the desired behavior. Which perspective does your implementation take? How could we implement the other approach?